stackslib/src/chainstate/stacks/boot/signers-voting.clar

```diff
@@ -62,35 +81,96 @@
+;;  Each signer votes for the aggregate public key for the next reward cycle.
+;;  This vote must happen after the list of signers has been set by the node,
+;;  which occurs in the first block of the prepare phase. The vote is concluded
+;;  when the threshold of `threshold-consensus / 1000` is reached for a
+;;  specific aggregate public key. The vote is weighted by the amount of
+;;  reward slots that the signer controls in the next reward cycle. The vote
+;;  may require multiple rounds to reach consensus, but once consensus is
+;;  reached, later rounds will be ignored.
+;;
+;;  Arguments:
+;;   * signer-index: the index of the calling signer in the signer set (from
+;;     `get-signers` in the .signers contract)
+;;   * key: the aggregate public key that this vote is in support of
+;;   * round: the voting round for which this vote is intended
+;;   * reward-cycle: the reward cycle for which this vote is intended
+;;  Returns:
+;;   * `(ok true)` if the vote was successful
+;;   * `(err <code>)` if the vote was not successful (see errors above)
+(define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint) (reward-cycle uint))
+    (let ((tally-key {reward-cycle: reward-cycle, round: round, aggregate-public-key: key})
+            ;; vote by signer weight
+            (signer-weight (try! (get-current-signer-weight signer-index)))
+            (new-total (+ signer-weight (default-to u0 (map-get? tally tally-key))))
+            (total-weight (try! (get-total-weight reward-cycle))))
+        ;; Check that the key has not yet been set for this reward cycle
+        (asserts! (is-none (map-get? aggregate-public-keys reward-cycle)) (err ERR_OUT_OF_VOTING_WINDOW))
+        ;; Check that the aggregate public key is the correct length
+        (asserts! (is-eq (len key) u33) (err ERR_ILL_FORMED_AGGREGATE_PUBLIC_KEY))
+        ;; Check that aggregate public key has not been used in a previous reward cycle
+        (asserts! (is-valid-aggregate-public-key key reward-cycle) (err ERR_DUPLICATE_AGGREGATE_PUBLIC_KEY))
+        ;; Check that signer hasn't voted in this reward-cycle & round
+        (asserts! (map-insert votes {reward-cycle: reward-cycle, round: round, signer: tx-sender} {aggregate-public-key: key, signer-weight: signer-weight}) (err ERR_DUPLICATE_VOTE))
+        ;; Check that the round is incremented by at most 1
+        (try! (update-last-round reward-cycle round))
+        ;; Update the tally for this aggregate public key candidate
+        (map-set tally tally-key new-total)
+        ;; Update used aggregate public keys
+        (map-set used-aggregate-public-keys key reward-cycle)
+        (print {
+            event: "voted",
+            signer: tx-sender,
+            reward-cycle: reward-cycle,
+            round: round,
+            key: key,
+            new-total: new-total,
+        })
+        ;; Check if consensus has been reached
+        (and
```

Comment:
I know `and` evaluates arguments left to right and will short-circuit, but can this be an `(if ...)` expression instead, just for better readability?

Reply:
I slightly prefer the `and`, but it shouldn't be a huge difference in cost, and after writing the `if` version, I do see that it could be better for readability.

Reply:
Refactored in 252a98cc9